### DISCUSSION TITLE
how change from  Lamda to Filter

### DISCUSSION THREAD

**Discussion by koch-ge** [dbuezas/lovelace-plotly-graph-card#245] [Q&A] [answered]
Hallo,
I copy'd a Lampda from matteobrusa for daly gasconsumption and it functions absoluteli super.
Now I have to change to Filter and Idon't know because I can no javascript.
I tryed it with -delta but the result is false after scrolling X-Axsis. with the Lamda it was perfect.
Perhubs can someone helb. Here my attempt.
```yaml
type: custom:plotly-graph
entities:
  - entity: sensor.gasincubicmeter
    name: täglicher Gasverbrauch
    show_value: true
    type: bar
    statistic: state
    period: day
    texttemplate: ' %{y:.2f} m³'
    filters:
      - fn: |-
          ({ys , xs}) => {  
            const DAY= 24*60*60*1000;
            let today= xs[0] - (xs[0].getTime() % DAY) 
                  +60*1000* new Date().getTimezoneOffset();
            const result = {x:[], y:[]}
            let last= parseFloat(ys[0]);
            let tomorrow= today + DAY;
            let firstday= today;
            let x,y
            for (let i = 0; i < xs.length; i+=1) {
              x = xs[i]
              y = parseFloat(ys[i]) 
              if (x>tomorrow){
                if (firstday != today) {
                  result.y.push(y-last)
                  result.x.push(new Date(today))
                }
                last = y
                today= tomorrow 
                tomorrow+= DAY
              }
            }   
            result.y.push(y-last)
            result.x.push(new Date(today))
            return result
          }
      - filter: i>0
refresh_interval: 10
hours_to_show: 4d
```

**Comment by dbuezas** [comment, marked as answer]
If I interpret the lambda correctly, this is computing the delta (or derivative in day units). Just do:
```yaml
filters:
  - delta
```

otherwise, just change `result.x` and `result.y` to `result.xs` and `result.ys`. And this too: `const result = {xs:[], ys:[]}`

**Comment by dbuezas** [comment]
If that's not enough let me know and I'll test/debug it myself. I suggest you also take a look at the [filters section of the readme](https://github.com/dbuezas/lovelace-plotly-graph-card#filters) you may get some ideas for this or other plots, these are fun to compose together :)

**Reply by koch-ge** [reply]
Thanks for your fast respond.
I have tried filter:-delta befor i wrote hier and the result was my problem.
-delta  shows the right values on initial show of the card. But if than I scroll the x-Axis or change the time to show on the x-Axis with buttons then the values are no longer correct they change the value.
on the Lambda I could scroll as I want and the values ar always correct.
If I use your suggestions from befor "otherwise, just change result.x and result.y to result.xs and result.ys. And this too: const result = {xs:[], ys:[]}"  it has the right value initial but not on scrolling. It is also exactly the same as with -delta
please excuse my very bad english.

**Reply by koch-ge** [reply]
![image](https://user-images.githubusercontent.com/114064428/217333170-f08aa3e3-15ac-406d-9701-c5b115599be4.png)
I had also tried this but also the same function and mailfunction like with -delta

**Reply by dbuezas** [reply]
Dein English ist eigentlich ziemlich gut, man verstehet alles ohne Probleme :) 
Right, another user found something about how HomeAssistant responds to queries on the statistics database: The way this card works is by taking the range of dates visible in the x-axis as [min_date, max_date], and it uses that to ask HomeAssistant for the statistics within that range. The issue is that when min_date is not right at the start of the day (let's say it is at mid day), then the daily statistics for that day only include half of the day.

Either that or I have introduced some weird bug.

Here are some options and experiments you can try:
1. Roll back to an older version of this card and check if the behaviour is the same. My assumption is that it will be the same, and that the old behaviour you describe is related to a home assistant version (or you just didn't notice it before). BUT if you find that the behaviour is different with an older version of this card, then I want to know about it so I try to find and fix the bug.
2. In order to avoid scrolling half days, use range buttons from [here](https://github.com/dbuezas/lovelace-plotly-graph-card/#range-selector-buttons). If that works for you I can help you tweak the buttons for zoom and scroll so they go in steps of full days.
3. Consider using `hours_to_show: current_week` (or current_month) which will ensure that the x axis starts exactly at midnight.

**Reply by koch-ge** [reply]
I found: the behaviour  with an older version of this card and "Lambda" is the same. So You are right it will be related to a home assistant update.

**Reply by dbuezas** [reply]
Ok, good to know. Then you'll have to go for option (2) and/or (3). Let me know if you need help customizing the buttons so they go in whole days.
